docs/kns_apps_tutorial.md

```diff
@@ -0,0 +1,347 @@
+
+While traditional Jarvice downstreams are made for HPC jobs and parallel short term jobs, KNS has been designed to host long running jobs with massive scaleup capabilities. KNS is in theory capable or running anything that can run on a Kubernetes cluster, making it a very nice shared "sandbox".
+
+KNS apps are docker images that contains files (mostly helm/kubectl/kustomize templates) and instructions to be used to deploy Kubernetes based apps, like Kubeflow, ArgoCD, Kubeai, etc. When launching a new job, the KNS will create a dedicated nested K3S cluster for the job, pull the app image from within this cluster, and use image's instructions to deploy the expected resources. Ingresses will allow user to access the nested cluster services.
+
+Note: some apps example for KNS can be found in the [Nimbix kns-apps repository](https://github.com/nimbix/kns-apps).
+
+## 2. The AppDef.json KNS key
+
+KNS introduces a new key, `nestedkubernetes`, inside traditional *AppDef.json* file of Jarvice applications, at `commands` level.
+
+A very basic hello world app would use the following AppDef.json file:
+
+```json
+{
+    "name": "Hello World",
+    "description": "Hello world app for KNS.",
+    "author": "oxedions",
+    "licensed": false,
+    "appdefversion": 2,
+    "classifications": [
+        "Uncategorized"
+    ],
+    "machines": [
+        "*"
+    ],
+    "vault-types": [
+        "FILE",
+        "BLOCK",
+        "BLOCK_ARRAY",
+        "OBJECT"
+    ],
+    "commands": {
+        "Hello": {
+            "path": "/usr/bin/true",
+            "name": "Start nested http server",
+            "description": "Start a nested http server for testing and to say hello world.",
+            "parameters": {},
+            "nestedkubernetes": {
+                "kubernetes_version": "",
+                "enable_gotty_shell": "true",
+                "ingress_type": "subdomain",
+                "targets": [
+                    {
+                        "service_name": "web",
+                        "service_port": 8080,
+                        "service_namespace": "default"
+                    }
+                ],
+                "templates_settings": {
+                    "global_retry": 20,
+                    "global_retry_sleep": 15
+                },
+                "templates": [
+                    {
+                        "name": "google-sample",
+                        "type": "kubectl",
+                        "path": "google-sample.yml"
+                    }
+                ]
+            }
+        }
+    },
+    "image": {
+        "type": "image/png",
+        "data": ""
+    }
+}
+```
+
+Lets examine this file step by steps. We will only focus on the commands part, everything else is the same than traditional Jarvice apps.
+
+```json
+    "commands": {
+        "Hello": {
+```
+
+* `commands` is a list, so you can add any app entry points as needed. This can be useful to create an all-in-one app image for example.
```

Comment:
```suggestion
* `commands` is a list, so you can add any app entry points as needed. This can be useful for creating an all-in-one app image for example.
```